src/tests/integration/MetaMorpho.t.sol

```diff
@@ -0,0 +1,300 @@
+    }
+
+    function testWithdrawLiquid(uint256 target1, uint256 target2) public {
+        uint256 marketSupplyStart = morpho.market(marketParams.id()).totalSupplyAssets;
+        
+        (,,, uint256 line,) = VatAbstract(vat).ilks(ilk);
+        vm.assume(target1 < line * 1e18 / 1e45);
+        
+        adjustDebt(int256(target1));
+        
+        vm.assume(target2 <= pool.assetBalance());
+        
+        uint256 marketSupplyMiddle = morpho.market(marketParams.id()).totalSupplyAssets;
+        
+        uint256 depositedAssets1 = min(target1, line * 1e18 / 1e45);
+        uint256 expectedWithdraw = min(pool.maxWithdraw(), pool.assetBalance() - target2);
+        uint256 depositedAssets2 = depositedAssets1 - expectedWithdraw;
+
+        // Set target assets at `target2` and exec.
+        vm.prank(operator);
+        plan.setTargetAssets(target2);
+        hub.exec(ilk);
+
+        uint256 marketSupplyEnd = morpho.market(marketParams.id()).totalSupplyAssets;
+        
+        assertGe(marketSupplyMiddle, marketSupplyStart);
+        assertLe(marketSupplyEnd, marketSupplyMiddle);
+        assertEq(marketSupplyEnd, marketSupplyMiddle - expectedWithdraw);
+        assertEq(marketSupplyEnd, marketSupplyStart + depositedAssets2);
+    }
+
+    function testWithdrawIlliquid(uint256 target1, uint256 target2, uint256 borrow) public {
+        uint256 marketSupplyStart = morpho.market(marketParams.id()).totalSupplyAssets;
+        uint256 marketBorrowStart = morpho.market(marketParams.id()).totalBorrowAssets;
+        vm.assume(borrow < marketSupplyStart - marketBorrowStart);
+        
+        (,,, uint256 line,) = VatAbstract(vat).ilks(ilk);
+        vm.assume(target1 < line * 1e18 / 1e45);
+        
+        adjustDebt(int256(target1));
+        adjustLiquidity(-int256(borrow));
+        
+        vm.assume(target2 <= pool.assetBalance());
+        
+        uint256 marketSupplyMiddle = morpho.market(marketParams.id()).totalSupplyAssets;
+        
+        uint256 depositedAssets1 = min(target1, line * 1e18 / 1e45);
+        uint256 expectedWithdraw = min(pool.maxWithdraw(), pool.assetBalance() - target2);
+        uint256 depositedAssets2 = depositedAssets1 - expectedWithdraw;
+
+        // Set target assets at `target2` and exec.
+        vm.prank(operator);
+        plan.setTargetAssets(target2);
+        hub.exec(ilk);
+
+        uint256 marketSupplyEnd = morpho.market(marketParams.id()).totalSupplyAssets;
+        
+        assertGe(marketSupplyMiddle, marketSupplyStart);
+        assertLe(marketSupplyEnd, marketSupplyMiddle);
+        assertEq(marketSupplyEnd, marketSupplyMiddle - expectedWithdraw);
+        assertEq(marketSupplyEnd, marketSupplyStart + depositedAssets2);
+    }
+}
+
+function min(uint256 x, uint256 y) pure returns (uint256) {
+    return x < y ? x : y;
+}
+function max(uint256 x, uint256 y) pure returns (uint256) {
+    return x > y ? x : y;
+}
```

Comment:
MetaMorpho imports `Math` (from OZ, through ERC4626), we could reuse that

Reply:
We are actually going to remove the dependencies before this gets merged, so best to leave this.